### ISSUE TITLE
why the Relative Ranking Positions S is not used?

### ISSUE THREAD

**Issue by luxiaogen** [Jo-wang/TCA#5]
ranking_positions = torch.zeros_like(indices)
            ranking_positions.scatter_(2, indices, torch.arange(N-1).unsqueeze(0).unsqueeze(1).repeat(B, self.n_head, 1).type_as(indices)) # 从“第 k 名是谁”变成了“某个 token 排第几   [1,12,196]
            # 得到了每个头中每个token对应的排名 ➡️ 算每个 token 在多个 attention head 里平均排第几名
            average_ranking_positions = ranking_positions.float().mean(dim=1) / (N - 1) # [1,196]

            cls_attn = cls_attn.mean(dim=1)  # [B, N-1]=[1,196] 多头注意力取平均

            if cache_attn_weights is not None: # 如果有缓存排名，再叠加 | 带一点跨层记忆
                average_ranking_positions = average_ranking_positions + 0.4 * cache_attn_weights
            # 最重要、直接保留 [1,196],177-2*20=137 ➡️ [1,137]
            # TODO cls_attn➡️average_ranking_positions
            _, idx = torch.topk(cls_attn, (left_tokens-2*merged_tokens), dim=1, largest=True, sorted=True)
            # idx1 top-left_tokens 的 token的索引 [1,177]
            _, idx1 = torch.topk(cls_attn, (left_tokens), dim=1, largest=True, sorted=True)
